### PR TITLE
fix: make GetWorkspacesEligibleForTransition return even less false positives

### PR DESCRIPTION
**Pull request by DanielleMaywood**
Relates to https://github.com/coder/coder/issues/15082

Further to https://github.com/coder/coder/pull/15429, this reduces the amount of false-positives returned by the 'is eligible for autostart' part of the query. We achieve this by calculating the 'next start at' time of the workspace, storing it in the database, and using it in our `GetWorkspacesEligibleForTransition` query.

The prior implementation of the 'is eligible for autostart' query would return _all_ workspaces that at some point in the future _might_ be eligible for autostart. This now ensures we only return workspaces that _should_ be eligible for autostart.

We also now pass `currentTick` instead of `t` to the `GetWorkspacesEligibleForTransition` query as otherwise we'll have one round of workspaces that are skipped by `isEligibleForTransition` due to `currentTick` being a truncated version of `t`.

### CODE REVIEW

File: coderd/provisionerdserver/provisionerdserver.go
Comment (johnstcn) [inline coderd/provisionerdserver/provisionerdserver.go:1456]: I don't think you can ignore the second return value ("allowed") here?

File: coderd/workspaces.go
Comment (johnstcn) [inline coderd/workspaces.go:886]: What if the template schedule store does not allow this next start time?

File: coderd/database/queries/workspaces.sql
Comment (DanielleMaywood) [reply coderd/database/queries/workspaces.sql:638]: Oops, yes my bad, happy to revert this back

File: coderd/provisionerdserver/provisionerdserver.go
Comment (DanielleMaywood) [reply coderd/provisionerdserver/provisionerdserver.go:1456]: Looking at it again maybe I slightly misunderstood it the first time round. 

So looking at it: `NextAutostart` returns the next autostart time after `at` (in this case the value `now`), and returns if this lands on a valid day of the week.

I guess this could cause false-positives to occur on days where autostart is not valid? 

Maybe more important to note that I've just spotted whilst re-reading it: the time returned by `NextAutostart` is localised to the schedules timezone instead of UTC. I think we will probably need to `.UTC()` on the returned time.

File: coderd/workspaces.go
Comment (DanielleMaywood) [reply coderd/workspaces.go:886]: In the case that it is disallowed we'll end up returning a false-positive until eventually it is allowed again. This is unfortunately not ideal. 

A potential resolution for this was discussed and we'll probably have a `NextAllowedAutostart` function in a future PR that will return the first _allowed_ autostart.

**Comment by dannykopping** [comment]
_Removed myself as reviewer; Cian & Mathias have much more context here._

File: coderd/database/migrations/000277_workspace_next_start_at.up.sql
Comment (johnstcn) [inline coderd/database/migrations/000277_workspace_next_start_at.up.sql:1]: Obligatory reminder to double-check migration number before merge 👍

File: coderd/database/dbmem/dbmem.go
Comment (mafredri) [inline coderd/database/dbmem/dbmem.go:6957]: Suggestion: Can we can add a note here that if we check zero so that we can update it and that it does't mean the workspace is actually eligible for transition?

**Review by mafredri — COMMENTED** [review]
I think this approach looks fine on the surface level, but I'm a bit worried that I'm not seeing any invalidation for "next start at". I.e. changes that mean we should disregard or recompute it.

For instance, it could be set in the future, but a change is done to start it sooner. Will it propagate?

Just a few thoughts that pop into my mind:

- What happens if a template scheduling options are changed?
- What happens if workspace settings are updated (e.g. autostart schedule)?
- What happens if a workspace is updated to a new template version
- What happens if a user is disabled after the nextStartAt has been set (this might be OK, just mentioning for completeness)

File: coderd/database/migrations/000277_workspace_next_start_at.up.sql
Comment (aaronlehmann) [inline coderd/database/migrations/000277_workspace_next_start_at.up.sql:1]: Should there be an index on this colum so `GetWorkspacesEligibleForTransition` doesn't require a sequential scan?

**Comment by DanielleMaywood** [comment]
> What happens if a template scheduling options are changed?
> What happens if a workspace is updated to a new template version

It appears I completely forgot to consider the template being updated, my bad, will resolve this! Thanks for spotting 👍  

> What happens if workspace settings are updated (e.g. autostart schedule)?

I've handled this (as far as I can tell), the `/workspaces/{workspace}/autostart [put]` endpoint updates the `next_start_at` column.

File: enterprise/coderd/schedule/template.go
Comment (DanielleMaywood) [reply enterprise/coderd/schedule/template.go:318]: Just to check I understand. You're suggesting we have a `TRIGGER` on certain conditions (those that invalidate our `next_start_at`) so we can keep `next_start_at` in sync? 

As the logic for calculating the `next_start_at` is easier to run on `coderd` - due to a lack of builtin support for cron expressions in Postgres - I assume this `TRIGGER` would `NOTIFY` our `coderd` server to perform this update?

My understanding (and please correct me if I'm wrong) is that `NOTIFY` doesn't _guarantee_ delivery, so could it be possible that we fail to update a `next_start_at` with this approach? If so, is the likelihood of this small enough to not be a concern?

Comment (mafredri) [reply enterprise/coderd/schedule/template.go:355]: (Side note: One risk in the trigger approach is that unless our relationships between tables is super clear (i.e. require one vs multiple joins), it could add a lot of overhead to updating the tables that have triggers.)

Comment (DanielleMaywood) [reply enterprise/coderd/schedule/template.go:355]: > This would force the Go code to re-fetch the row on next iteration and recompute the new start?

Yeah this is definitely an option we could look at doing

Comment (Emyrk) [reply enterprise/coderd/schedule/template.go:355]: I forgot to respond yesterday :facepalm: 

@mafredri yes this is correct. Just using `TRIGGER` to invalidate the column (set to null).

At the moment we add the invalidation to the http routes that affect it. That just feels a bit leaky and might be a bug down the road when we forget to invalidate it.

Comment (DanielleMaywood) [reply enterprise/coderd/schedule/template.go:358]: I've got this working!

When `autostart_schedule` is updated but `next_start_at` _isn't_ updated, the database sets `next_start_at` to `NULL`. On the next run of the lifecycle executor, it sees that the column is `NULL`, so it computes it and updates it in the database.

```
coder=# select name, autostart_schedule, next_start_at from workspaces;
     name     |    autostart_schedule    |     next_start_at
--------------+--------------------------+------------------------
 teal-bass-93 | CRON_TZ=UTC 0 13 * * 1-5 | 2024-11-28 13:00:00+00
(1 row)

coder=# update workspaces set autostart_schedule='CRON_TZ=UTC 30 14 * * 1-5' where name='teal-bass-93';
UPDATE 1

coder=# select name, autostart_schedule, next_start_at from workspaces;
     name     |    autostart_schedule     | next_start_at
--------------+---------------------------+---------------
 teal-bass-93 | CRON_TZ=UTC 30 14 * * 1-5 |
(1 row)

... a minute later ...

coder=# select name, autostart_schedule, next_start_at from workspaces;
     name     |    autostart_schedule     |     next_start_at
--------------+---------------------------+------------------------
 teal-bass-93 | CRON_TZ=UTC 30 14 * * 1-5 | 2024-11-28 14:30:00+00
(1 row)
```

File: coderd/autobuild/lifecycle_executor.go
Comment (DanielleMaywood) [inline coderd/autobuild/lifecycle_executor.go:145]: In our lifecycle executor we use `currentTick` for checking eligibility, so it makes sense to also use it for our database query.

File: coderd/database/migrations/000278_workspace_next_start_at.up.sql
Comment (mafredri) [inline coderd/database/migrations/000278_workspace_next_start_at.up.sql:22]: I'm thinking we also need a trigger for changes to a template time-til-dormant? Or do I misunderstand that interaction?

My thought process: if the time-til-dormant is reduced, we might've previously calculated a next start for the workspace which might no longer be valid because dormancy was sped up.

It's probably a fringe edge-case, but might be hard to debug customer reports if they run into it.

I'm thinking this applies to template allowed days changing as well. Tomorrow might've been blocked but is now unblocked yet we computed the autostart for the day after.

I'm wondering if there are other such interactions between template and workspace we should cover as well?

File: coderd/database/queries/workspaces.sql
Comment (mafredri) [inline coderd/database/queries/workspaces.sql:641]: Let's mention the trigger too here.

File: coderd/database/queries/workspaces.sql
Comment (mafredri) [inline coderd/database/queries/workspaces.sql:802]: We should probably make a (partial) index for this.

```sql
CREATE INDEX ON workspaces (template_id) WHERE deleted = FALSE;
```

We can do it in a follow up PR though. We don't actually have a single index on this column so it'll be good to have, especially since it's a foreign key and one would expect fk lookups to happen.

File: coderd/schedule/autostart.go
Comment (mafredri) [inline coderd/schedule/autostart.go:39]:
```suggestion
	// Our autostart day requirements may block the day, so to ensure we've exhausted all
```
Maybe this is entirely wrong, but it's what I would expect the comment to be as I'm having trouble understanding it. Otherwise what does cron schedules/weekly basis mean?

File: enterprise/coderd/schedule/template.go
Comment (mafredri) [inline enterprise/coderd/schedule/template.go:35]: ❤️

**Review by mafredri — COMMENTED** [review]
Nice work! Seeing it, I'm not sure how I feel about the db trigger approach, but I like the enforcement and think it's worth giving it a shot. I think we still have some possibilities for data inconsistency without a trigger on the templates table, though.

File: coderd/database/migrations/000278_workspace_next_start_at.up.sql
Comment (DanielleMaywood) [reply coderd/database/migrations/000278_workspace_next_start_at.up.sql:22]: I think I agree with you here, I'll try and think if there are any other similar interactions 👍

File: coderd/schedule/autostart.go
Comment (DanielleMaywood) [reply coderd/schedule/autostart.go:39]: I can see how my comment is poorly worded. 

What I was trying to say is that since the cron schedule is parsed with `cron.Weekly` it is scoped to a single-week (so you cant specify first Monday of the month for example). So the basis of the comment is saying we only need to check the next 7 days to have fully exhausted the set of possible autostarts.

File: coderd/database/dbmem/dbmem.go
Comment (DanielleMaywood) [reply coderd/database/dbmem/dbmem.go:7003]: I don't _think_ it'd cause flakes, but instead cause dbmem tests to fail when postgres tests do not. Happy to add it though

Comment (DanielleMaywood) [reply coderd/database/dbmem/dbmem.go:7003]: On second thought, I've checked `dbmem.go` for how we emulate triggers and I'm not sure there is a good way to do it.

File: coderd/schedule/autostart.go
Comment (mafredri) [reply coderd/schedule/autostart.go:39]: Gotcha, that makes sense, thanks for clarifying 👍🏻.

File: coderd/database/migrations/000278_workspace_next_start_at.up.sql
Comment (DanielleMaywood) [reply coderd/database/migrations/000278_workspace_next_start_at.up.sql]: > I'm thinking we also need a trigger for changes to a template time-til-dormant? Or do I misunderstand that interaction?

> My thought process: if the time-til-dormant is reduced, we might've previously calculated a next start for the workspace which might no longer be valid because dormancy was sped up.


After digging into this a little bit more, I don't think this is an issue.

I had forgotten to include `dormant_at IS NOT NULL` in `autostart` part of the `GetWorkspacesEligibleForTransition` SQL query, so I've added that now (it already existed in the coderd lifecycle executor side).

Instead of handling changes to `time_til_dormancy`, I've instead modified the existing trigger to check if the workspace is dormant. When a workspace is marked as dormant, the workspace's `next_start_at` is made NULL.

> I'm thinking this applies to template allowed days changing as well. Tomorrow might've been blocked but is now unblocked yet we computed the autostart for the day after.

This edge case has been handled in the `EnterpriseTemplateScheduleStore`

File: enterprise/coderd/schedule/template.go
Comment (johnstcn) [reply enterprise/coderd/schedule/template.go:358]: I think this will work nicely!

File: coderd/database/dbmem/dbmem.go
Comment (DanielleMaywood) [reply coderd/database/dbmem/dbmem.go:7003]: Ah that's unfortunate. 

I think attempting to emulate this trigger isn't worth it then. If someone writes a test that is dependent on the trigger (which they ideally shouldn't) then they'll have to disable it when running under `dbmem.go`.

File: coderd/database/queries/workspaces.sql
Comment (DanielleMaywood) [reply coderd/database/queries/workspaces.sql:806]: Have added this now 👍 It's a very small change so I'm happy for it to be included.

**Review by johnstcn — APPROVED** [review]
Approving to unblock.

**Pull request merged**
